kv/kvclient/rangefeed: TestDBClientScan failed

kv/kvclient/rangefeed.TestDBClientScan [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12483859?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12483859?buildTab=artifacts#/) on master @ [354b78c2b4f43d0499d34c32e012a076c30a40bb](https://github.com/cockroachdb/cockroach/commits/354b78c2b4f43d0499d34c32e012a076c30a40bb):


```
=== RUN   TestDBClientScan
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/8a396ce5ffb24698ad44a501ceca79c0/logTestDBClientScan1652679289
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:165: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    conditional_wrap.go:190: 
        pkg/kv/kvclient/rangefeed_test/pkg/kv/kvclient/rangefeed/db_adapter_external_test.go:63: (TestDBClientScan)
        	NOTICE: .SplitRange() called via implicit interface StorageLayerInterface;
        HINT: consider using .StorageLayer().SplitRange() instead.
    db_adapter_external_test.go:277: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8a396ce5ffb24698ad44a501ceca79c0/logTestDBClientScan1652679289
--- FAIL: TestDBClientScan (61.88s)
=== RUN   TestDBClientScan/parallel_scan_requests
    db_adapter_external_test.go:165: completed scan for /Tenant/10/Table/104/{1-2}
    db_adapter_external_test.go:173: condition failed to evaluate within 45s: from db_adapter_external_test.go:175: still waiting for barrier (1/3)
    --- FAIL: TestDBClientScan/parallel_scan_requests (46.32s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #113570 kv/kvclient/rangefeed: TestDBClientScan failed [A-testing C-bug C-test-failure O-robot T-kv-replication branch-release-23.2]
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDBClientScan.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/rangefeed.TestDBClientScan [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13733731?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13733731?buildTab=artifacts#/) on master @ [e46153e9571b4fb2a6053ac2ddc4d08bd35575a6](https://github.com/cockroachdb/cockroach/commits/e46153e9571b4fb2a6053ac2ddc4d08bd35575a6):


```
=== RUN   TestDBClientScan
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/8a396ce5ffb24698ad44a501ceca79c0/logTestDBClientScan1796789676
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    conditional_wrap.go:190: 
        pkg/kv/kvclient/rangefeed_test/pkg/kv/kvclient/rangefeed/db_adapter_external_test.go:63: (TestDBClientScan)
        	NOTICE: .SplitRange() called via implicit interface StorageLayerInterface;
        HINT: consider using .StorageLayer().SplitRange() instead.
    db_adapter_external_test.go:277: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8a396ce5ffb24698ad44a501ceca79c0/logTestDBClientScan1796789676
--- FAIL: TestDBClientScan (61.23s)
=== RUN   TestDBClientScan/parallel_scan_requests
    db_adapter_external_test.go:165: completed scan for /Tenant/10/Table/104/1{-/250}
    db_adapter_external_test.go:165: completed scan for /Tenant/10/Table/104/{1/250-2}
    db_adapter_external_test.go:173: condition failed to evaluate within 45s: from db_adapter_external_test.go:175: still waiting for barrier (2/3)
    --- FAIL: TestDBClientScan/parallel_scan_requests (46.15s)
```

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #116540 kv/kvclient/rangefeed: TestDBClientScan failed [C-bug C-test-failure O-robot P-2 T-kv-replication X-duplicate branch-release-23.2.0-rc]
- #113570 kv/kvclient/rangefeed: TestDBClientScan failed [A-testing C-bug C-test-failure O-robot P-2 T-kv-replication branch-release-23.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDBClientScan.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/rangefeed.TestDBClientScan [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13764677?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13764677?buildTab=artifacts#/) on master @ [9a98c70638ce481e119aba8cc850ff44dc7c69e7](https://github.com/cockroachdb/cockroach/commits/9a98c70638ce481e119aba8cc850ff44dc7c69e7):


```
=== RUN   TestDBClientScan
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/8a396ce5ffb24698ad44a501ceca79c0/logTestDBClientScan2482216310
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    conditional_wrap.go:190: 
        pkg/kv/kvclient/rangefeed_test/pkg/kv/kvclient/rangefeed/db_adapter_external_test.go:63: (TestDBClientScan)
        	NOTICE: .SplitRange() called via implicit interface StorageLayerInterface;
        HINT: consider using .StorageLayer().SplitRange() instead.
    db_adapter_external_test.go:277: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8a396ce5ffb24698ad44a501ceca79c0/logTestDBClientScan2482216310
--- FAIL: TestDBClientScan (62.13s)
=== RUN   TestDBClientScan/parallel_scan_requests
    db_adapter_external_test.go:165: completed scan for /Tenant/10/Table/104/{1-2}
    db_adapter_external_test.go:173: condition failed to evaluate within 45s: from db_adapter_external_test.go:175: still waiting for barrier (1/3)
    --- FAIL: TestDBClientScan/parallel_scan_requests (46.51s)
```

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #116540 kv/kvclient/rangefeed: TestDBClientScan failed [C-bug C-test-failure O-robot P-2 T-kv-replication X-duplicate branch-release-23.2.0-rc]
- #113570 kv/kvclient/rangefeed: TestDBClientScan failed [A-testing C-bug C-test-failure O-robot P-2 T-kv-replication branch-release-23.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDBClientScan.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvclient/rangefeed/db_adapter_external_test.go
+++ b/pkg/kv/kvclient/rangefeed/db_adapter_external_test.go
@@ -17,6 +17,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
@@ -60,7 +61,7 @@ func TestDBClientScan(t *testing.T) {
 	beforeAny := db.Clock().Now()
 
 	scratchKey := append(ts.Codec().TenantPrefix(), keys.ScratchRangeMin...)
-	_, _, err := srv.SplitRange(scratchKey)
+	_, _, err := srv.StorageLayer().SplitRange(scratchKey)
 	require.NoError(t, err)
 
 	mkKey := func(k string) roachpb.Key {
@@ -143,13 +144,36 @@ func TestDBClientScan(t *testing.T) {
 			db, ts.Codec(), "defaultdb", "foo")
 		fooSpan := fooDesc.PrimaryIndexSpan(ts.Codec())
 
-		// Refresh the DistSender range cache. ScanWithOptions will split the scan
-		// requests itself based on the range cache and assert on that split, before
-		// sending the requests to the DistSender.
-		_, err := db.Scan(ctx, fooSpan.Key, fooSpan.EndKey, 0)
-		require.NoError(t, err)
+		// Ensure the splits make it into the meta ranges and range cache. Simply
+		// running a DistSender scan does not appear sufficient in rare cases.
+		//
+		// ScanWithOptions will split the scan requests itself based on the range
+		// cache and assert on that split, before sending them to the DistSender.
+		ds := ts.DistSenderI().(*kvcoord.DistSender)
+		testutils.SucceedsSoon(t, func() error {
+			// Flush the cache, and repopulate it via a scan. This looks at the
+			// canonical range descriptors rather than the possibly stale meta ranges.
+			ds.RangeDescriptorCache().Clear()
+			_, err := db.Scan(ctx, fooSpan.Key, fooSpan.EndKey, 0)
+			require.NoError(t, err)
 
-		// We expect 4 splits -- we'll start the scan with parallelism set to 3.
+			var descs []roachpb.RangeDescriptor
+			iter := kvcoord.MakeRangeIterator(ds)
+			for iter.Seek(ctx, roachpb.RKey(fooSpan.Key), kvcoord.Ascending); iter.Valid(); iter.Next(ctx) {
+				desc := iter.Desc()
+				if !fooSpan.Overlaps(desc.KeySpan().AsRawSpanWithNoLocals()) {
+					break
+				}
+				descs = append(descs, *desc)
+			}
+			if len(descs) == 4 {
+				t.Logf("range cache has 4 ranges: %v", descs)
+				return nil
+			}
+			return errors.Errorf("range cache has %d ranges: %v", len(descs), descs)
+		})
+
+		// We have 4 ranges -- we'll start the scan with parallelism set to 3.
 		// We will block these scans from completion until we know that we have 3
 		// concurrently running scan requests.
 		var parallelism = 3

